IMU Driver Implementation

rakoon1208 opened the pull request:
- Created an IMU Driver for the IIM-20670 based off its respective datasheet and implemented it into Phoenix
- Created documentation for the IMU driver

rakoon1208 committed:
feat: add IMU driver for IIM-20670 along with documentation for driver specs

rakoon1208 committed:
refactor: adjusted madgwick-test to use the new API for the IMU driver; added new tests

rakoon1208 committed:
feat: implemented IMU driver

n123xyz commented on crates/common-arm/src/drivers/iim20670.rs:
```diff
@@ -0,0 +1,773 @@
+    CS: OutputPin<Error = CSE>,
+    DELAY: DelayUs<u32>,
+{
+    /// Create driver and initialize with default configuration
```
use a fn default() implementation if you want a default constructor.

rakoon1208 replied:
I didn't want a default constructor but its supposed to be a constructor with default settings (its meant to stay as a custom constructor since apparently the SPI peripheral, GPIO pin, and delay provider cant have a default)

n123xyz commented on crates/common-arm/src/drivers/iim20670.rs:
```diff
@@ -0,0 +1,773 @@
+    }
+
+    /// Create driver with validation checks
+    pub fn new_with_validation(spi: SPI, mut cs: CS, mut delay: DELAY) -> Result<Self, Error<SPIE, CSE>> {
```
Move this to be fn new(spi: SPI, mut cs: CS, mut delay: DELAY)

rakoon1208 committed:
fix: resolve build errors in pheonix

errors were caused by missing driver import in main.rs and missing float import in madgwick_service.rs, also some syntax errors

BLM16 commented on crates/common-arm/src/drivers/iim20670.rs:
```diff
@@ -0,0 +1,773 @@
+    
+    fn convert_temp(&self, raw: i16) -> f32 { 
+        25.0 + (raw as f32 / 20.0) 
+    }
```
Do these actually do conversions? It seems to me to be more of a "normalize" type method converting read values into meaningful ones. If the SDS calls this conversion, great. Just figured I'd ask since it doesn't really seem to be conversion. Correct me if I'm wrong.

rakoon1208 replied:
Yea you're right, they aren't really conversions; they are meant to be scaling/normalization functions that will convert the raw ADC values into meaningful/measurable units. I made changes to the comments to better reflect this.

BLM16 commented on phoenix/Cargo.toml:
```diff
@@ -23,6 +23,7 @@ panic-probe = { workspace = true }
 chrono = { workspace = true }
 messages = { workspace = true }
 madgwick = { workspace = true }
+m = "0.1.0"
```
I see `m` as a dependency in a few files. It should probably be made a workspace dependency in this case to sync versions.

BLM16 replied:
Although if you're just using it for testing and it's thus synced to the Madgwick tests, it's probably fine.

rakoon1208 replied:
Changed to a workplace dependency

BLM16 commented on phoenix/src/main.rs:
```diff
@@ -90,6 +90,22 @@ mod app {
+            stm32h7xx_hal::gpio::Pin<
+                'F',
+                6,
+                stm32h7xx_hal::gpio::Output<stm32h7xx_hal::gpio::PushPull>,
```
The new lines between generic types here is probably unnecessary. I'd just leave this as one line. Not really an issue though, just looks funny to me 🙃

rakoon1208 replied:
Yea, I did it this way since that's how the barometer types were structured, and since I didn't wanna touch those, I figured I'd just format mine the same way the barometer types were done. I can still change it regardless though.

BLM16 reviewed (commented):
Looks good overall! Just a few minor comments, plus Noah's review. Thanks for working on this!

rakoon1208 committed:
refactor: use workspace dependancy for m crate in madgwick-test

rakoon1208 committed:
refactor: rename data conversion functions to scaling/normalization functions and constructor functions, both for clarity
